extension/json/json_scan.cpp

```diff
@@ -0,0 +1,497 @@
+	buffer_offset += part2_size;
+
+	// We copied the object, so we are no longer reading the previous buffer
+	if (--previous_buffer_handle->readers == 0) {
+		current_reader->RemoveBuffer(current_buffer_handle->buffer_index - 1);
+	}
+
+	objects[0] = ParseLine((char *)reconstruct_ptr, part1_size + part2_size, lines[0], gstate.bind_data.ignore_errors);
+}
+
+void JSONScanLocalState::ReadUnstructured(idx_t &count) {
+	const auto max_obj_size = reconstruct_buffer.GetSize();
+	yyjson_read_err error;
+	for (; count < STANDARD_VECTOR_SIZE; count++) {
+		const auto obj_start = buffer_ptr + buffer_offset;
+		const auto obj_copy_start = buffer_copy_ptr + buffer_offset;
+
+		idx_t remaining = buffer_size - buffer_offset;
+		if (remaining == 0) {
+			break;
+		}
+
+		// Read next JSON doc
+		auto read_doc = JSONCommon::ReadDocumentUnsafe((char *)obj_start, remaining, JSONCommon::STOP_READ_FLAG,
+		                                               json_allocator.GetYYJSONAllocator(), &error);
+		if (error.code == YYJSON_READ_SUCCESS) {
+			idx_t line_size = yyjson_doc_get_read_size(read_doc);
+			lines[count] = JSONLine(obj_copy_start, line_size);
+			TrimWhitespace(lines[count]);
+
+			buffer_offset += line_size;
+			SkipWhitespace(buffer_ptr, buffer_offset, buffer_size);
+		} else if (error.pos > max_obj_size) {
+			JSONCommon::ThrowParseError(obj_copy_start, remaining, error,
+			                            "Have you tried increasing maximum_object_size?");
+		} else if (error.code == YYJSON_READ_ERROR_UNEXPECTED_END && !is_last) {
+			// Copy remaining to reconstruct_buffer
+			const auto reconstruct_ptr = reconstruct_buffer.get();
+			memcpy(reconstruct_ptr, obj_copy_start, remaining);
+			prev_buffer_remainder = remaining;
+			buffer_offset = buffer_size;
+			break;
+		} else {
+			JSONCommon::ThrowParseError(obj_copy_start, remaining, error);
+		}
+		objects[count] = read_doc;
+	}
+}
+
+void JSONScanLocalState::ReadNewlineDelimited(idx_t &count, const bool &ignore_errors) {
+	for (; count < STANDARD_VECTOR_SIZE; count++) {
+		auto line_start = buffer_ptr + buffer_offset;
+		idx_t remaining = buffer_size - buffer_offset;
+		if (remaining == 0) {
+			break;
+		}
+
+		// Search for newline
+		auto line_end = NextNewline(line_start, remaining);
+
+		if (line_end == nullptr) {
+			// We reached the end of the buffer
+			if (!is_last) {
+				// Last bit of data belongs to the next batch
+				buffer_offset = buffer_size;
+				break;
+			}
+			line_end = line_start + remaining;
+		}
+		idx_t line_size = line_end - line_start;
+
+		objects[count] = ParseLine((char *)line_start, line_size, lines[count], ignore_errors);
+
+		buffer_offset += line_size;
+		SkipWhitespace(buffer_ptr, buffer_offset, buffer_size);
+	}
+}
+
+idx_t JSONScanLocalState::GetBatchIndex() const {
+	// Is this enough? Needs to be lower than PipelineBuildState::BATCH_INCREMENT = 10000000000000
+	return current_reader->file_index * 1000000 + current_buffer_handle->buffer_index;
```

Comment:
Do we need to do the `1000000 * file_index` trick? Can't we know how many batch indexes are in each file while we are reading?

For example, if we have a 100MB file, and read 1MB buffers, we know there are 100 batch indexes in that file.

Reply:
What about compression? If the original 100MB file is 10MB compressed, we will still read 100 batch indexes, but then we have to _estimate_ the compression ratio to estimate how many indexes there are in the file.

Reply:
That's fair - in the CSV reader we dynamically assign batch indexes instead to avoid having to do this. Perhaps the same is possible for the JSON reader.

Reply:
I thought this was tricky, but I've implemented dynamic batch indices properly now